#ERC analysis of plastid-nuclear coevolution in angiosperms

Included here are scripts that perform the phylogenetic components of our pipeline for evolutionary rate covariation (ERC) analyses in angiosperms.

This pipeline was used to perform the analyses in the following study:

*Forsythe ES, Williams AM, Sloan DB, Genome-wide signatures of plastid-nuclear coevolution point to repeated perturbations of plastid proteostasis systems across angiosperms (in revision).*

Available from: https://doi.org/10.1101/2020.08.28.272872

Note that this code is provided as a means to reproduce the results of the above manuscript and as a starting point for those wishing to perform similar analyses in different taxa but the pipeline has not yet been developed into a stand-alone tool.

##Contents of this repository
This repository contains:

+ *<An R script for parsing large numbers of nuclear gene trees to make them suitable for ERC analyses>*

+ *<A python script for pulling out specific sequences from fasta files (called from within the R script)>*


